### PR TITLE
fix: support multimodal OpenAI-compatible embeddings

### PR DESCRIPTION
**Pull request by TatshSiow**
## Description

Adds image embedding support for multimodal models configured through the OpenAI-compatible provider. PipesHub now sends image inputs using vLLM's multimodal `messages` extension instead of dropping them as an unsupported provider.

The text embedding path is unchanged. Image requests use the configured endpoint, model, and API key, with bounded concurrency and per-image failure handling so an unsupported or failed image request does not interrupt text indexing.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (Is there an addition/change in API Request, Response? If yes, documentation is required)
- [ ] Performance improvement
- [ ] Code refactoring
- [ ] Security fix

## Related Issues

fix issue #2869 

## How Has This Been Tested?

The change was deployed against a self-hosted vLLM endpoint serving `Qwen/Qwen3-VL-Embedding-8B`. Text and image requests both returned 4,096-dimensional vectors through `/v1/embeddings`.

The existing Qdrant collection was recreated and a mixed set of PDF, PPTX, DOCX, and XLSX records was reindexed. PipesHub stored native image vectors without unsupported-provider warnings or image-embedding failures.

Focused unit tests cover provider dispatch, endpoint construction, authorization, multimodal payload shape, successful responses, and graceful request failure. The full unit suite was not run because the production image does not include the test extras.

### Test Configuration

- [ ] Unit tests pass
- [ ] Integration tests pass
- [x] Manual testing completed
- [ ] Cross-browser testing (if applicable)
- [ ] Mobile responsiveness tested (if applicable)

## Core Functionality Testing

Please confirm that the following core functionalities are working as expected:

- [ ] **Search capabilities** are working correctly
- [ ] **Knowledge search** is functioning properly
- [x] **Connector indexing** is working as expected
- [ ] **Citations** are displaying and linking correctly
- [ ] **Documentation** is updated at https://docs.pipeshub.com/introduction

## What You Have Tested

Please describe what you have specifically tested:

- [x] Feature works in development environment
- [ ] Feature works in staging environment
- [x] Error handling scenarios tested
- [x] Edge cases considered and tested
- [x] Performance impact assessed
- [x] Security implications reviewed

### Test Results

- OpenAI-compatible embedding health check returned HTTP 200.
- Text embeddings returned 4,096-dimensional vectors.
- Image embeddings returned 4,096-dimensional vectors.
- Qdrant collection was recreated with a 4,096-dimensional dense vector configuration.
- Reindexing stored 334 native image vectors across 10 image-store operations.
- No OpenAI-compatible image embedding failures were logged.
- Rebuilt PipesHub container remained healthy.

## Screenshots/Videos

Not applicable; this change has no UI behavior.

**Note:** Screenshots are required before merge for UI changes.

## Code Quality Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Documentation

- [ ] Documentation has been updated (if applicable)
- [ ] API documentation updated (if applicable)
- [ ] README updated (if applicable)
- [ ] Changelog updated (if applicable)

No public PipesHub API changes are introduced.

## Security Considerations

- [x] No sensitive data is exposed
- [x] Input validation is implemented where necessary
- [x] Authentication/authorization is properly handled
- [x] No security vulnerabilities introduced

The request uses the existing configured API key and sends it as a bearer token only when present.

## Breaking Changes

- [ ] This PR introduces breaking changes
- [ ] Migration guide provided (if applicable)
- [ ] Version bump required

No breaking changes.

## Performance Impact

- [ ] No performance impact
- [ ] Performance improved
- [x] Performance impact assessed and acceptable

Image embedding adds one request per image. Requests share one asynchronous client and are bounded by the existing concurrency limit of five.

## Dependencies

- [x] No new dependencies added
- [ ] New dependencies are necessary and approved
- [ ] Dependencies updated and tested

## Deployment Notes

- The configured OpenAI-compatible endpoint must support vLLM's multimodal `messages` extension on `/v1/embeddings`.
- Configure the endpoint as the API base URL, for example `http://host:port/v1`, not the full `/v1/embeddings` route.
- Enable multimodal embedding for the model configuration.
- Recreate the vector collection and reindex existing documents when changing embedding models or dimensions.

## Checklist Before Merge

- [ ] All tests are passing
- [ ] Code review completed and approved
- [ ] Documentation updated and reviewed
- [x] Screenshots/videos attached for UI changes (not applicable)
- [x] Core functionality verified
- [x] Security review completed (if applicable)
- [x] Performance impact assessed
- [ ] Ready for production deployment

## Additional Notes

This PR intentionally contains only the OpenAI-compatible image embedding path and focused tests. It does not modify the existing text embedding implementation or other providers.

---

**For Reviewers:**

Please pay special attention to:

- The vLLM-specific `messages` payload used for image embeddings
- Graceful behavior when an OpenAI-compatible server does not support multimodal embedding
- Preservation of the existing text embedding path

## Summary by CodeRabbit

* **New Features**
  * Added support for generating image embeddings through OpenAI-compatible services.
  * Supports configurable endpoints, authentication, and compatible image request formats.
  * Successfully generated embeddings are converted into searchable vector points.
  * Agent creation responses now indicate whether default models are being used.

* **Bug Fixes**
  * Invalid images, failed requests, and unavailable services are safely skipped.
  * Improved embedding context-length handling across compatible providers.

### CODE REVIEW

File: backend/python/app/modules/transformers/vectorstore.py [hunk 1]
```diff
@@ -871,6 +875,81 @@ async def process_batch(
...
+                                    ],
+                                }
+                            ],
+                            "encoding_format": "float",
+                        },
+                    )
+                    response.raise_for_status()
+                    embedding = response.json()["data"][0]["embedding"]
+                    if not isinstance(embedding, list) or not embedding:
+                        raise ValueError("response contains no embedding vector")
```
[the remark this answers is not in the document]
Comment (TatshSiow) [reply backend/python/app/modules/transformers/vectorstore.py]: Addressed in 926cefd39. The response is now validated before nested access: payload object, non-empty data list, first-item object, and non-empty numeric embedding vector. Added parameterized coverage for seven malformed payload shapes; the focused OpenAI-compatible tests pass (7/7).

File: backend/python/app/modules/transformers/vectorstore.py [hunk 2]
```diff
@@ -895,6 +974,10 @@ async def _process_image_embeddings(
             return await self._process_image_embeddings_bedrock(image_chunks, image_base64s)
         elif self.embedding_provider == EmbeddingProvider.JINA_AI.value:
             return await self._process_image_embeddings_jina(image_chunks, image_base64s)
+        elif self.embedding_provider == EmbeddingProvider.OPENAI_COMPATIBLE.value:
+            return await self._process_image_embeddings_openai_compatible(
+                image_chunks, image_base64s
+            )
```
[the remark this answers is not in the document]
Comment (TatshSiow) [reply backend/python/app/modules/transformers/vectorstore.py]: Addressed in 926cefd39 with a local provider-to-processor registry in `_process_image_embeddings`, preserving the existing processor signatures and unsupported-provider behavior without introducing a separate factory abstraction. The complete image-embedding test group passes (25/25).

File: backend/python/tests/unit/modules/transformers/test_vectorstore.py [hunk 3]
```diff
@@ -2248,6 +2261,90 @@ async def test_jina_invalid_images_filtered(self):
...
+        "payload",
+        [
+            [],
+            {},
+            {"data": []},
+            {"data": [None]},
+            {"data": [{}]},
+            {"data": [{"embedding": []}]},
+            {"data": [{"embedding": [0.1, "invalid"]}]},
+        ],
```
[the remark this answers is not in the document]
Comment (TatshSiow) [reply backend/python/tests/unit/modules/transformers/test_vectorstore.py:2367]: Addressed in 011e55348. Boolean values are now explicitly excluded from numeric embedding coordinates, and `[True]` was added to the malformed-response regression cases. Focused OpenAI-compatible tests pass (8/8).

**Comment by TatshSiow** [comment]
This is fix of vLLM or other OpenAI Compatible endpoints to correctly support multimodal embeddings

File: backend/python/tests/unit/modules/transformers/test_vectorstore.py [hunk 4]
```diff
@@ -2248,6 +2261,126 @@ async def test_jina_invalid_images_filtered(self):
...
+            points = await vs._process_image_embeddings_openai_compatible(
+                [{"metadata": {"page": 1}, "image_uri": "aW1hZ2U="}],
+                ["aW1hZ2U="],
+            )
+
+        assert len(points) == 1
+        assert mock_client.post.await_count == 2
+        second_call = mock_client.post.await_args_list[1]
+        content = second_call.kwargs["json"]["messages"][0]["content"]
+        assert content[0]["image_url"]["url"] == "data:image/jpeg;base64,aW1hZ2U="
```
[the remark this answers is not in the document]
Comment (kaushalwaghela-art) [reply backend/python/tests/unit/modules/transformers/test_vectorstore.py:2331]: fixed

File: backend/python/tests/unit/utils/test_openrouter_integration.py [hunk 5]
```diff
@@ -98,6 +98,16 @@ def test_dimensions_forwarded(self, mock_cls):
...
 
+    @patch("langchain_openai.embeddings.OpenAIEmbeddings")
+    def test_sends_plain_strings_not_token_arrays(self, mock_cls):
+        """OpenRouter proxies non-OpenAI models that 400 on token-ID input."""
+        mock_cls.return_value = MagicMock()
+        get_embedding_model(
+            EmbeddingProvider.OPENROUTER.value,
+            _config("nvidia/llama-nemotron-embed-vl-1b-v2:free"),
+        )
+        assert mock_cls.call_args.kwargs["check_embedding_ctx_length"] is False
```
[the remark this answers is not in the document]
Comment (kaushalwaghela-art) [reply backend/python/tests/unit/utils/test_openrouter_integration.py:109]: fixed

**Review by abhishek-pipeshub — APPROVED** [review]
